src/scatter.jl

```diff
@@ -0,0 +1,189 @@
+export scatter!, scatter
+
+## Scatter API
+#   - Scatter:
+#     - scatter(op, src, idx)
+#     - scatter!(op, dst, src, idx)
+#   - Scatter destination backpropagation
+#     - ∇scatter_dst!
+#   - Scatter source backpropagation
+#     - ∇scatter_src
+#     - ∇scatter_src!
+#
+
+function _check_dims(Ndst, Nsrc, N, Nidx)
+    @assert Ndst - N == Nsrc - Nidx
+    dims = Ndst - N
+    if dims < 0
+        throw(ArgumentError("dims must be non-negative but got dims=$dims"))
+    end
+    return dims
+end
+
+_check_input(idx::AbstractArray{<:Integer}, arr) = checkbounds(arr, minimum(idx):maximum(idx))
+
+function _check_input(idx::AbstractArray{<:Tuple}, arr)
+    pairs = map(xs -> Base.OneTo(maximum(xs)), zip(idx...))
+    checkbounds(arr, pairs...)
+end
+
+function _check_output(idx::AbstractArray{<:IntOrTuple}, arr, src, dims)
+    pre_dims = axes(src)[1:dims]
+    post_dims = Base.OneTo.(maximum_dims(idx))
+    checkbounds(arr, pre_dims..., post_dims...)
+end
+
+
+"""
+    scatter!(op, dst, src, idx)
+
+Scatter operation, which scatters data in `src` and assigns to `dst` according to `idx`.
+With the data going to the same place, specified operation is applied on to reduce data.
+For each index `k` in `idx`, accumulate values in `dst` according to
+
+    dst[idx[k]...] = (op).(dst[idx[k]...], src[k...])
+
+# Arguments
+- `op`: operations to be applied on `dst` and `src`, e.g. `+`, `-`, `*`, `/`, `max` and `min`.
+- `dst`: the destination for `src` to aggregate to. This argument will be mutated.
+- `src`: the source data for aggregating.
+- `idx`: the mapping for aggregation from source (index) to destination (value).
+The index of `idx` is corresponding to the index of `src` and the value of `idx` is
+corresponding to the index of `dst`. The value of `idx` can be `Int` or `Tuple` type.
+
+    dst[:, ..., idx[k]...] = (op).(dst[:, ..., idx[k]...], src[:, ..., k...])
+"""
+function scatter!(op,
+                  dst::AbstractArray{Tdst,Ndst},
+                  src::AbstractArray{Tsrc,Nsrc},
+                  idx::AbstractArray{<:Integer,Nidx}) where {Tdst<:Real,Tsrc<:Real,Ndst,Nsrc,Nidx}
```

Comment:
```suggestion
                  idx::AbstractArray{<:Integer,Nidx}) where {Tdst,Tsrc,Ndst,Nsrc,Nidx}
```
this restriction is not needed, we can support scatter acting on non-numerical arrays

Reply:
Done.